python/dgl/dataloading/labor_sampler.py

```diff
@@ -0,0 +1,176 @@
+        optimization steps that many times.
+    layer_dependency : bool, default ``False`` 
+        Specifies whether different layers should use same random variates.
+    batch_dependency : int, default ``1``
+        Makes it so that batch i and batch i + batch_dependency are independent but
+        consecutive batches are dependent. The bigger this parameter, the more dependency between
+        batches.
+    prefetch_node_feats : list[str] or dict[ntype, list[str]], optional
+        The source node data to prefetch for the first MFG, corresponding to the
+        input node features necessary for the first GNN layer.
+    prefetch_labels : list[str] or dict[ntype, list[str]], optional
+        The destination node data to prefetch for the last MFG, corresponding to
+        the node labels of the minibatch.
+    prefetch_edge_feats : list[str] or dict[etype, list[str]], optional
+        The edge data names to prefetch for all the MFGs, corresponding to the
+        edge features necessary for all GNN layers.
+    output_device : device, optional
+        The device of the output subgraphs or MFGs.  Default is the same as the
+        minibatch of seed nodes.
+
+    Examples
+    --------
+    **Node classification**
+
+    To train a 3-layer GNN for node classification on a set of nodes ``train_nid`` on
+    a homogeneous graph where each node takes messages from 5, 10, 15 neighbors for
+    the first, second, and third layer respectively (assuming the backend is PyTorch):
+
+    >>> sampler = dgl.dataloading.LaborSampler([5, 10, 15])
+    >>> dataloader = dgl.dataloading.DataLoader(
+    ...     g, train_nid, sampler,
+    ...     batch_size=1024, shuffle=True, drop_last=False, num_workers=4)
+    >>> for input_nodes, output_nodes, blocks in dataloader:
+    ...     train_on(blocks)
+
+    If training on a heterogeneous graph and you want different number of neighbors for each
+    edge type, one should instead provide a list of dicts.  Each dict would specify the
+    number of neighbors to pick per edge type.
+
+    >>> sampler = dgl.dataloading.LaborSampler([
+    ...     {('user', 'follows', 'user'): 5,
+    ...      ('user', 'plays', 'game'): 4,
+    ...      ('game', 'played-by', 'user'): 3}] * 3)
+
+    If you would like non-uniform neighbor sampling:
+
+    >>> g.edata['p'] = torch.rand(g.num_edges())   # any non-negative 1D vector works
+    >>> sampler = dgl.dataloading.NeighborSampler([5, 10, 15], prob='p')
+
+    **Edge classification and link prediction**
+
+    This class can also work for edge classification and link prediction together
+    with :func:`as_edge_prediction_sampler`.
+
+    >>> sampler = dgl.dataloading.LaborSampler([5, 10, 15])
+    >>> sampler = dgl.dataloading.as_edge_prediction_sampler(sampler)
+    >>> dataloader = dgl.dataloading.DataLoader(
+    ...     g, train_eid, sampler,
+    ...     batch_size=1024, shuffle=True, drop_last=False, num_workers=4)
```

Comment:
Did you test with `as_edge_prediction_sampler`?

Reply:
Yes, I did. It seems to work. I don't have tests written for those scenarios though.